src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -68,7 +70,21 @@ private record LinkRequest(FunctionDescriptor descriptor, LinkerOptions options)
     private final SoftReferenceCache<LinkRequest, UpcallStubFactory> UPCALL_CACHE = new SoftReferenceCache<>();
 
     @Override
-    public MethodHandle downcallHandle(FunctionDescriptor function, Option... options) {
+    @CallerSensitive
+    public final MethodHandle downcallHandle(MemorySegment symbol, FunctionDescriptor function, Option... options) {
+        downcallHandle0(Reflection.getCallerClass(), function, options);
+        SharedUtils.checkSymbol(symbol);
+        return downcallHandle(function, options).bindTo(symbol);
+    }
```

Comment:
This doesn't look quite right, as `downcallHandle` is being called twice? (forgot to push?).

I suggest calling `Reflection::ensureNativeAccess` immediately in each overload, and then delegating to `downcallHandle0` which doesn't check the access.

Reply:
Just a mistake on my part - fixed now